### PR TITLE
RES-998

### PR DESCRIPTION
**Pull request by mvacaporale**
Bug fixes and porting of tests from nupic.torch.

### CODE REVIEW

File: nupic/tensorflow/layers/k_winners.py [hunk 1]
```diff
@@ -401,7 +422,7 @@ def update_duty_cycle(self, inputs):
 
         """
         batch_size = tf.shape(inputs)[0]
-        self.learning_iterations += batch_size
+        self.learning_iterations.assign(self.learning_iterations + batch_size)
```
Comment (lscheinkman) [inline nupic/tensorflow/layers/k_winners.py]: It should be `assign_add`:
```python
self.learning_iterations.assign_add(batch_size)
```

File: tests/k_winner_test.py [hunk 2]
```diff
@@ -38,63 +43,350 @@
...
+
             testing_utils.layer_test(KWinners,
                                      kwargs=kwargs,
-                                     input_data=self.x,
-                                     expected_output=expected)
+                                     input_data=x,
+                                     expected_output=expected,
+                                     )
 
+    def test_two(self):
```
Comment (lscheinkman) [inline tests/k_winner_test.py]: missing decorator for "all modes"
```python
@keras_parameterized.run_all_keras_modes
```

File: tests/k_winner_test.py [hunk 3]
```diff
@@ -38,63 +43,350 @@
...
+        kwargs = {
+            "percent_on": 0.333,
+            "k_inference_factor": 1.5,
+            "boost_strength": 1.0,
+            "boost_strength_factor": 0.5,
+            "duty_cycle_period": 1000,
+        }
+
+        # Test layer output in and out of training mode.
+        with context.eager_mode():
```
Comment (lscheinkman) [inline tests/k_winner_test.py]: why is this block running only in "eager_mode"?

File: tests/k_winner_test.py [hunk 4]
```diff
@@ -38,63 +43,350 @@
...
+            expected = np.zeros(x.shape, dtype=np.float32)
+            expected[0, 1] = 1.0
+            expected[0, 4] = 1.0
+            expected[1, 1] = 1.0
+            expected[1, 4] = 1.0
+
+            result = kw(x, training=True)
+            self.assertAllEqual(result, expected)
+
+    def test_three(self):
```
Comment (lscheinkman) [inline tests/k_winner_test.py]: missing decorator for "all modes"
```python
@keras_parameterized.run_all_keras_modes
```

File: tests/k_winner_test.py [hunk 5]
```diff
@@ -38,63 +43,350 @@
...
+
+        kwargs = {
+            "percent_on": 0.333,
+            "k_inference_factor": 1.5,
+            "boost_strength": 1.0,
+            "boost_strength_factor": 0.5,
+            "duty_cycle_period": 1000,
+        }
+
+        with context.eager_mode():
```
Comment (lscheinkman) [inline tests/k_winner_test.py]: why is this block running only in "eager_mode"?

File: tests/k_winner_test.py [hunk 6]
```diff
@@ -189,6 +486,43 @@ def test_four(self):
...
+
+        kwargs = {
+            "percent_on": 0.25,
+            "k_inference_factor": 0.5,
+            "boost_strength": 1.0,
+            "boost_strength_factor": 0.5,
+            "duty_cycle_period": 1000,
+        }
+
+        with context.eager_mode():
```
Comment (lscheinkman) [inline tests/k_winner_test.py]: why is this block running only in "eager_mode"?

File: tests/k_winner_test.py [hunk 7]
```diff
@@ -189,6 +486,43 @@ def test_four(self):
                                      input_data=x,
                                      expected_output=expected)
 
+    def test_five(self):
```
Comment (lscheinkman) [inline tests/k_winner_test.py]: missing decorator for "all modes"
```python
@keras_parameterized.run_all_keras_modes
```

File: nupic/tensorflow/layers/k_winners.py [hunk 8]
```diff
@@ -77,6 +77,7 @@ def compute_kwinners(x, k, duty_cycles, boost_strength):
     :return:
         A tensor representing the activity of x after k-winner take all.
     """
+    x = tf.cast(x, tf.float32)
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/tensorflow/layers/k_winners.py]: Good question. I don't think this is needed. I may have thought it was when I was fixing the type mismatch errors on Friday.

File: nupic/tensorflow/layers/k_winners.py [hunk 9]
```diff
@@ -150,7 +151,6 @@ def __init__(
         self.percent_on = percent_on
         self.percent_on_inference = percent_on * k_inference_factor
         self.k_inference_factor = k_inference_factor
-        self.learning_iterations = 0
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/tensorflow/layers/k_winners.py]: Will do. What's the motivation for this change? 
Is it for clarity?
I don't believe it's required by the keras.Layer class.

File: nupic/tensorflow/layers/__init__.py [hunk 10]
```diff
@@ -18,4 +18,4 @@
 #
 # http://numenta.org/licenses/
 # ----------------------------------------------------------------------
-from .k_winners import KWinners, KWinners2d
+from .k_winners import KWinners, KWinners2d, compute_kwinners
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/tensorflow/layers/__init__.py]: Good point.

File: nupic/tensorflow/layers/k_winners.py [hunk 11]
```diff
@@ -303,10 +318,10 @@ def get_config(self):
     def update_duty_cycle(self, x):
         batch_size = tf.shape(x)[0]
         duty_cycles_shape = self.duty_cycles.shape
-        self.learning_iterations += batch_size
+        self.learning_iterations.assign(tf.add(self.learning_iterations, batch_size))
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/tensorflow/layers/k_winners.py]: Fair

File: nupic/tensorflow/layers/k_winners.py (hunk 1 above)
Comment (mvacaporale) [reply nupic/tensorflow/layers/k_winners.py]: Fair

File: nupic/tensorflow/layers/k_winners.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/tensorflow/layers/k_winners.py]: It is good practice to declare all instance variables in the in `__init__`

**Pull request merged**